### ISSUE TITLE
Investigate `Publish Post-Signing Assets` failure - trying to publish assets that already exist

### ISSUE THREAD

**Issue by MilenaHristova** [dotnet/arcade-services#2418]
Publishing fails when a previous version of the asset with the same name already exists in the feed

**Build**: https://dev.azure.com/dnceng/internal/_build/results?buildId=2055126&view=logs&j=30cb8153-2b4a-535b-a3dc-8e694bdfc9a9&t=b32cda20-409d-5a6b-f8e9-19fb1861da8e&l=3549

Steps:
- [ ] https://github.com/dotnet/arcade-services/issues/2621

**Error**
```
Asset 'D:\a\_work\1\a\upload\BlobArtifacts\158197\dotnet-sdk-source-6.0.112.tar.gz' already exists with different contents at 'Sdk/6.0.112-servicing.22578.14/dotnet-sdk-source-6.0.112.tar.gz'
```

**Goal**
Investigate when that problem occurs and implement a fix

**Comment by epananth** [comment]
@mmitche not sure why this is happening. Isn't staging pipeline intelligent enough not to have this error? Also its only this stage.

**Comment by MilenaHristova** [comment]
@epananth I had mistaken the stage, it's in `Publish Post-Signing Assets` but I think it happens whenever there is a retry. 
The Arcade task that's publishing checks if the hash of the blob content is same as the hash of the file and fails if the contents differ
https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Build.Tasks.Feed/src/AzureStorageAssetPublisher.cs#L37-L51
But they shouldn't differ because nothing changes with the assets, they are downloaded again, copied to the same folders and published so we can either
- overwrite the packages when publishing 
- change the error to a info or warning (now the error doesn't interrupt the publishing so maybe it shouldn't be shown as error)

**Comment by mmitche** [comment]
I don't think the package can be overwritten by default. Those storage containers should be immutable. In .NET 6, it's quite possible that the content **did** change because of post-build signing. In those cases, the storage container would need to have its immutability removed and the file manually overwritten. Are you seeing this behavior in releases other than .NET 6?

**Comment by MilenaHristova** [comment]
@mmitche this is .NET 7  https://dev.azure.com/dnceng/internal/_build/results?buildId=2168911&view=logs&j=06e47235-ea4e-5696-bf7a-60141ec0f646&t=23030f86-a850-5629-e37b-e34af5164200
But if there is a rerun then how can the content change between attempts?

**Comment by mmitche** [comment]
@MilenaHristova Ahh dang. That would happen because Linux signing was rerun on the same input assets, leading to different content.

**Comment by MilenaHristova** [comment]
> @MilenaHristova Ahh dang. That would happen because Linux signing was rerun on the same input assets, leading to different content.

o ok I see, so the whole pipeline for .NET 7 servicing was ran twice, it published the first time, then the second run assets changed a bit because of signing and tried to publish again. In this case how important is it to have the new version of the asset in the storage?

**Comment by MilenaHristova** [comment]
@mmitche i think this will be fixed by the pipeline split that Epsi is working on? Then the signing will be run only once

**Comment by mmitche** [comment]
I took another look at this. I think it may be simpler than expected. I **think** all that needs to happen here is that a new storage container needs to be created if one already exists. In that step, we create feeds, checksum, and non-checksum asset containers. The default name for the various containers and feeds is based on the build number of the highest version SDK. For the feeds, we appear to append a suffix (-1, -2, -3 etc.) if the feed already exists, so we don't get conflicts. This does not appear to be the case for the containers. We just use the same name again and again.

If I remember right, the purpose of this particular job is just to make signed assets available _and_ linked via aka.ms (so you can dotnet-install with -Quality signed. The outputs here are not used downstream in the release pipeline, because these are not validated yet. The Publish CTI Validated Assets stage is responsible for that. That particular stage is supposed to write to a stable, predictable location and conflicts would be expected (and should be errors) if we rebuilt after we had signed off on a particular build was already signed off.

What is a little confusing is that there are actually two stages that publish signed assets. One is right after required validation, and the other one is right after SBOM generation. They publish to slightly different locations for installers. Maybe we should combine these.

Side note: One improvement that should be made is the stages/jobs should have better names that describe the purpose of the various publishing jobs.

**Comment by MilenaHristova** [comment]
@andriipatsula I will look at what's the difference between `Publish Post-Signing Assets` and `Publish Signed Assets` and how can we combine them. On first glance they do very similar things but `Publish Signed Assets` separates artifacts into assets, packages, symbols and NuGet packages and publishes them in a different way while `Publish Post-Signing Assets` uses Arcades `PublishArtifactsInManifest` to publish everything

**Comment by MilenaHristova** [comment]
I created another issue to combine the singing stages https://github.com/dotnet/arcade-services/issues/2621

**Comment by MilenaHristova** [comment]
This is done with https://dev.azure.com/dnceng/internal/_git/dotnet-release/pullrequest/32261
